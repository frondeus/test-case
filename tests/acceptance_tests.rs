#![cfg(test)]

mod acceptance {
    use insta::with_settings;
    use itertools::Itertools;
    use std::env;
    use std::path::PathBuf;
    use std::process::{Command, Output};

    fn get_snapshot_directory() -> String {
        PathBuf::from("snapshots")
            .join(env::var("SNAPSHOT_DIR").unwrap_or_else(|_| "rust-stable".to_string()))
            .to_str()
            .unwrap()
            .to_string()
    }

    fn retrieve_stdout(output: &Output) -> String {
        String::from_utf8_lossy(&output.stdout)
            .to_string()
            .lines()
            .filter(|s| !s.is_empty())
            .map(|line| match line.find("; finished in") {
                Some(idx) => &line[0..idx],
                None => line,
            })
            .sorted()
            .join("\n")
    }

    fn retrieve_stderr(output: &Output) -> String {
        String::from_utf8_lossy(&output.stderr)
            .to_string()
            .lines()
            .filter(|s| !s.is_empty())
            .map(|line| match line.find("; finished in") {
                Some(idx) => &line[0..idx],
                None => line,
            })
            .sorted()
            .join("\n")
    }

    #[test]
    fn basic() {
        with_settings!({snapshot_path => get_snapshot_directory()}, {
            let output = Command::new("cargo")
                .current_dir(PathBuf::from("acceptance_tests").join("basic"))
                .args(&["test"])
                .output()
                .expect("cargo command failed to start");

            let lines = retrieve_stdout(&output);
            let lines = lines
                .lines()
                .filter(|line| !line.contains("note"))
                .join("\n");
            insta::assert_display_snapshot!(lines);
        });
    }

    #[test]
    fn async_tests() {
        with_settings!({snapshot_path => get_snapshot_directory()}, {
            let output = Command::new("cargo")
                .current_dir(PathBuf::from("acceptance_tests").join("async_tests"))
                .args(&["test"])
                .output()
                .expect("cargo command failed to start");

            let lines = retrieve_stdout(&output);
            let lines = lines
                .lines()
                .filter(|line| !line.contains("note"))
                .join("\n");
            insta::assert_display_snapshot!(lines);
        });
    }

    #[test]
    fn test_item_reuse() {
        with_settings!({snapshot_path => get_snapshot_directory()}, {
            let output = Command::new("cargo")
                .current_dir(PathBuf::from("acceptance_tests").join("test_item_reuse"))
                .args(&["test"])
                .output()
                .expect("cargo command failed to start");

            let lines = retrieve_stdout(&output);
            let lines = lines
                .lines()
                .filter(|line| !line.contains("note"))
                .join("\n");
            insta::assert_display_snapshot!(lines);
        });
    }

    #[test]
    fn test_item_reuse_run() {
        with_settings!({snapshot_path => get_snapshot_directory()}, {
            let output = Command::new("cargo")
                .current_dir(PathBuf::from("acceptance_tests").join("test_item_reuse"))
                .args(&["run"])
                .output()
                .expect("cargo command failed to start");

            let mut lines = retrieve_stdout(&output);
            lines.push_str(&retrieve_stderr(&output));
            let lines = lines
                .lines()
                .filter(|line| !line.contains("waiting") && !line.contains("Finished") && !line.contains("Compiling"))
                .join("\n");

            insta::assert_display_snapshot!(lines);
        });
    }

    #[test]
    fn test_result() {
        with_settings!({snapshot_path => get_snapshot_directory()}, {
            let output = Command::new("cargo")
                .current_dir(PathBuf::from("acceptance_tests").join("result"))
                .args(&["test"])
                .output()
                .expect("cargo command failed to start");

            let lines = retrieve_stdout(&output);
            let lines = lines
                .lines()
                .filter(|line| !line.contains("note"))
                .join("\n");
            insta::assert_display_snapshot!(lines);
        });
    }
}
