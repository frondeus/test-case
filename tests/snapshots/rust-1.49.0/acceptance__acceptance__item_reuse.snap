---
source: tests/acceptance_tests.rs
assertion_line: 94
expression: lines

---
    internal_tested_function1::_3_expects_6
  left: `6`,
 right: `0`', src/lib.rs:8:1
---- internal_tested_function1::_3_expects_6 stdout ----
failures:
failures:
running 6 tests
test internal_tested_function1::_2_expects_4 ... ok
test internal_tested_function1::_3_expects_6 ... FAILED
test internal_tested_function2::_1_expects_0 ... ok
test internal_tested_function3::_1_expects_matching_3 ... ok
test internal_tested_function3::_2_expects_inconclusive6 ... ignored
test internal_tested_function4::_2_expects_panicking_some_can_t_ ... ok
test result: FAILED. 4 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out
thread 'internal_tested_function1::_3_expects_6' panicked at 'assertion failed: `(left == right)`
